_posts/2016-03-28-Mutual-Problems.md

```diff
@@ -0,0 +1,130 @@
+    attribution:
+        text: Image by Sean MacEntee (CC BY 2.0)
+        href: https://www.flickr.com/photos/smemon/15944989872/in/photostream/
+tags:
+- Java
+- HTTPS
+- Mutual Authentication
+--- 
+
+The HTTPS protocol is the well-established standard for securing our internet connections. Understanding how this protocol works is not problem and the corresponding [RFC document](https://tools.ietf.org/html/rfc2818) is available since 2000.
+
+Despite HTTPS is used so widely, you can still find a software which doesn't handle this protocol without unnecessary complexity. Unfortunately I'v experienced  problems during the implementation of [the mutual authentication](https://en.wikipedia.org/wiki/Mutual_authentication) in the language which should not surprise me at all. It is **Java**.
+
+## How does HTTPS work?
+
+Before I describe what kind of problems I've got with my implementation, I will describe how the mutual authentication works. The HTTPS protocol use the TLS/SSL protocol to secure the connection. The TLS/SSL protocol defines the authentication handshake which allows to connect any client with the server in the secure way. 
+During [The handshake](https://en.wikipedia.org/wiki/Transport_Layer_Security#TLS_handshake) the following steps are performed:
+
+- A client sends the message which initiates the connection.
+- The server sends its certificate to the client.
+- The client verifies the certificate using the certificate issued by the trusted authority.
+- The server sends the request for the client's certificate.
+- The client sends its certificate to the server.  
+- The server verifies the client's certificate.
+- The server and the client exchange the master secret which is used during the encryption of data.
+- Connection is established.  
+
+Combining my knowledge about the TLS/SSL handshake and my experience from the manual test with `curl` I've assumed that I'm going to need to just three files to implement the client side: **a client's certificate**, **a client's private key** and **a trusted certificate to verify the server's certificate**.
+
+## Java - the problem, the solution and why is it so hard
+
+Because it is quite unusual to use the mutual authentication every day, together with my teammate we asked for the assistance. The first look at the results served by *uncle Google* didn't revealed the complexity behind the implementation, but each click on the results led us to the more and more confusing solutions (some of them where from 90's). To make matters worse we had to use [Apache HttpComponents](https://hc.apache.org/) to implement our connection, but the most of the proposed solutions based on the pure **Java** libraries.  
+
+The knowledge from the internet allows us to establish:
+
+- **Java** cannot use directly any certificates or private keys (like e.g. `curl`)
+- **Java** requires separate files (**Java Keystores**) which can contain original certificates and keys. 
```

Comment:
It would be too easy otherwise ;-)

On some java versions, you also cannot import x509 certificates that contain text metadata before the base64 content. I lost several hours of my life on this :smile: